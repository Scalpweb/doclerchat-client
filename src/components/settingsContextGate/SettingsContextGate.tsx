import React, {useState, useEffect, useRef, useContext} from 'react';
import {TranslationContext} from 'components/translationContextGate/TranslationContextGate';

export interface Settings {
  username: string;
  interfaceColor: 'light' | 'dark';
  clockDisplay: '12' | '24';
  enableControlEnter: 'on' | 'off';
  language: string;
}

export const defaultSettings: Settings = {
  username: 'John Doe',
  interfaceColor: 'light',
  clockDisplay: '12',
  enableControlEnter: 'off',
  language: 'en'
};

export interface SettingsContextProps {
  fetch: () => Settings;
  store: (settings: Settings) => void;
  storeOne: (key: string, value: string | boolean) => void;
}

export const SettingsContext = React.createContext<SettingsContextProps>({
  fetch: () => defaultSettings,
  store: () => {
    return;
  },
  storeOne: () => {
    return;
  }
});

const STORAGE_KEY = 'doclerChatSettings';

interface Props {
  children: (settings: Settings) => React.ReactNode;
}

const fetchSettings = () => {
  const savedSettings = window.localStorage.getItem(STORAGE_KEY);
  if (savedSettings) {
    return JSON.parse(savedSettings);
  }
  return defaultSettings;
};

const SettingsContextGate = ({children}: Props) => {
  const translation = useRef(useContext(TranslationContext));
  const [settings, setSettings] = useState<Settings>(fetchSettings());

  const storeSettings = (newSettings: Settings) => {
    window.localStorage.setItem(STORAGE_KEY, JSON.stringify(newSettings));
    setSettings({...newSettings});
  };

  const storeOneSettings = (key: string, value: string) => {
    storeSettings({...fetchSettings(), [key]: value});
  };

  useEffect(() => {
    translation.current.setLanguage(fetchSettings()['language']);
  }, []);

  return (
    <SettingsContext.Provider
      value={{fetch: fetchSettings, store: storeSettings, storeOne: storeOneSettings}}>
      {children(settings)}
    </SettingsContext.Provider>
  );
};

export default SettingsContextGate;
