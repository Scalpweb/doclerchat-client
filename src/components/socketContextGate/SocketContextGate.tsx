import React, {useEffect, useState, useRef, useContext} from 'react';
import {SettingsContext} from 'components/settingsContextGate/SettingsContextGate';
import SocketManager, {Messages} from '../../utils/socket';
import Loader from 'components/loader/Loader';
import {makeTabBlink} from '../../utils/utils';

export interface SocketContextProps {
  send: (type: string, json: string) => void;
}

export const SocketContext = React.createContext<SocketContextProps>({
  send: () => {
    return;
  }
});

interface Props {
  children: (messages: Messages[], unread: number, resetUnreadCount: () => void) => React.ReactNode;
}

const SocketContextGate = ({children}: Props) => {
  const context = useRef(useContext(SettingsContext));

  const [connected, setConnected] = useState(false);
  const [messages, setMessages] = useState<Messages[]>([]);
  const [unreadCount, setUnreadCount] = useState(0);

  const update = () => {
    context.current.storeOne('username', SocketManager.getInstance().getUsername());
    setMessages([...SocketManager.getInstance().getMessages()]);
    setConnected(SocketManager.getInstance().isConnected());
    setUnreadCount((u) => u + 1);
    makeTabBlink();
  };

  const sendMessage = (type: string, data: string) => {
    SocketManager.getInstance().send(type, data);
  };

  useEffect(() => {
    SocketManager.getInstance().init(context.current.fetch().username);
    SocketManager.getInstance().addListener(update);
    update();
    return () => {
      SocketManager.getInstance().removeListener(update);
    };
  }, []);

  const resetUnreadCount = () => {
    setUnreadCount(0);
  };

  if (!connected) {
    return <Loader />;
  }

  return (
    <SocketContext.Provider value={{send: sendMessage}}>
      {children(messages, unreadCount, resetUnreadCount)}
    </SocketContext.Provider>
  );
};

export default SocketContextGate;
