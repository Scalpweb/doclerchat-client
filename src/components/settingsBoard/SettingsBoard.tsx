import React, {useRef, useContext, useEffect} from 'react';
import {
  SettingsContext,
  Settings,
  defaultSettings
} from 'components/settingsContextGate/SettingsContextGate';
import styles from './SettingsBoard.module.scss';
import {SocketContext} from 'components/socketContextGate/SocketContextGate';
import {TranslationContext, t} from 'components/translationContextGate/TranslationContextGate';

interface Props {
  settings: Settings;
}

const SettingsBoard = ({settings}: Props) => {
  const translation = useRef(useContext(TranslationContext));
  const context = useRef(useContext(SettingsContext));
  const socket = useRef(useContext(SocketContext));

  const updateSettings = (e: React.ChangeEvent<HTMLInputElement | HTMLSelectElement>) => {
    const fieldName = e.currentTarget.name;
    const value = e.currentTarget.value;
    context.current.store({
      ...settings,
      [fieldName]: value
    });
    if (fieldName === 'language') {
      translation.current.setLanguage(value);
    }
  };

  const updateUsername = (e: React.FocusEvent<HTMLInputElement>) => {
    context.current.storeOne('username', e.currentTarget.value);
    socket.current.send('username', e.currentTarget.value);
  };

  const resetDefaults = (e: React.MouseEvent<HTMLButtonElement>) => {
    e.preventDefault();
    e.stopPropagation();
    context.current.store(defaultSettings);
  };

  return (
    <div className={styles.container}>
      <div className={styles.form}>
        <div className={styles.field}>
          <label>{t('username')}</label>
          <input
            name="username"
            onChange={updateSettings}
            onBlur={updateUsername}
            type="text"
            value={settings.username}
          />
        </div>
        <div className={styles.field}>
          <label>{t('interfaceColor')}</label>
          <div className={styles.radio}>
            <input
              name="interfaceColor"
              onChange={updateSettings}
              type="radio"
              value="light"
              checked={settings.interfaceColor === 'light'}
            />{' '}
            {t('light')}
          </div>
          <div className={styles.radio}>
            <input
              name="interfaceColor"
              onChange={updateSettings}
              type="radio"
              value="dark"
              checked={settings.interfaceColor === 'dark'}
            />{' '}
            {t('dark')}
          </div>
        </div>
        <div className={styles.field}>
          <label>{t('clockDisplay')}</label>
          <div className={styles.radio}>
            <input
              name="clockDisplay"
              onChange={updateSettings}
              type="radio"
              value="12"
              checked={settings.clockDisplay === '12'}
            />{' '}
            12
          </div>
          <div className={styles.radio}>
            <input
              name="clockDisplay"
              onChange={updateSettings}
              type="radio"
              value="24"
              checked={settings.clockDisplay === '24'}
            />{' '}
            24
          </div>
        </div>
        <div className={styles.field}>
          <label>{t('sendOnCtrl')}</label>
          <div className={styles.radio}>
            <input
              name="enableControlEnter"
              onChange={updateSettings}
              type="radio"
              value="on"
              checked={settings.enableControlEnter === 'on'}
            />{' '}
            {t('on')}
          </div>
          <div className={styles.radio}>
            <input
              name="enableControlEnter"
              onChange={updateSettings}
              type="radio"
              value="off"
              checked={settings.enableControlEnter === 'off'}
            />{' '}
            {t('off')}
          </div>
        </div>
        <div className={styles.field}>
          <label>{t('language')}</label>
          <select name="language" onChange={updateSettings} value={settings.language}>
            <option value="fr">Fran√ßais</option>
            <option value="en">English</option>
          </select>
        </div>
        <div className={styles.field}>
          <button onClick={resetDefaults}>{t('resetDefault')}</button>
        </div>
      </div>
    </div>
  );
};

export default SettingsBoard;
