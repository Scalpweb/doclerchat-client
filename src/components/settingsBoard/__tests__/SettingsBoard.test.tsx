import React from 'react';
import {mount} from 'enzyme';
import {SocketContext} from '../../socketContextGate/SocketContextGate';
import {SettingsContext, defaultSettings} from '../../settingsContextGate/SettingsContextGate';
import SettingsBoard from '../SettingsBoard';

const settingsWrapper = {fetch: jest.fn(), store: jest.fn(), storeOne: jest.fn()};

describe('SettingsBoards', () => {
  it('update storage', () => {
    const component = mount(
      <SettingsContext.Provider value={settingsWrapper}>
        <SocketContext.Provider value={{send: jest.fn()}}>
          <SettingsBoard settings={defaultSettings} />
        </SocketContext.Provider>
      </SettingsContext.Provider>
    );
    expect(component).toMatchSnapshot();

    const stub = jest.spyOn(settingsWrapper, 'store');

    component
      .find('input')
      .at(1)
      .simulate('change', {currentTarget: {value: 'light'}});
    expect(stub).toHaveBeenCalledTimes(1);

    component
      .find('input')
      .at(2)
      .simulate('change', {currentTarget: {value: 'dark'}});
    expect(stub).toHaveBeenCalledTimes(2);
  });
});
