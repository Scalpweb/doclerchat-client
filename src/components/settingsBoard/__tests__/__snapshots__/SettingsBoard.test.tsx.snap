// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsBoards update storage 1`] = `
<SettingsBoard
  settings={
    Object {
      "clockDisplay": "12",
      "enableControlEnter": "off",
      "interfaceColor": "light",
      "language": "en",
      "username": "John Doe",
    }
  }
>
  <div>
    <div>
      <div>
        <label>
          Username
        </label>
        <input
          name="username"
          onBlur={[Function]}
          onChange={[Function]}
          type="text"
          value="John Doe"
        />
      </div>
      <div>
        <label>
          Interface color
        </label>
        <div>
          <input
            checked={true}
            name="interfaceColor"
            onChange={[Function]}
            type="radio"
            value="light"
          />
           
          Light
        </div>
        <div>
          <input
            checked={false}
            name="interfaceColor"
            onChange={[Function]}
            type="radio"
            value="dark"
          />
           
          Dark
        </div>
      </div>
      <div>
        <label>
          Clock display
        </label>
        <div>
          <input
            checked={true}
            name="clockDisplay"
            onChange={[Function]}
            type="radio"
            value="12"
          />
           
          12
        </div>
        <div>
          <input
            checked={false}
            name="clockDisplay"
            onChange={[Function]}
            type="radio"
            value="24"
          />
           
          24
        </div>
      </div>
      <div>
        <label>
          Send message on Ctrl+Enter
        </label>
        <div>
          <input
            checked={false}
            name="enableControlEnter"
            onChange={[Function]}
            type="radio"
            value="on"
          />
           
          On
        </div>
        <div>
          <input
            checked={true}
            name="enableControlEnter"
            onChange={[Function]}
            type="radio"
            value="off"
          />
           
          Off
        </div>
      </div>
      <div>
        <label>
          Language
        </label>
        <select
          name="language"
          onChange={[Function]}
          value="en"
        >
          <option
            value="fr"
          >
            Fran√ßais
          </option>
          <option
            value="en"
          >
            English
          </option>
        </select>
      </div>
      <div>
        <button
          onClick={[Function]}
        >
          Reset Defaults
        </button>
      </div>
    </div>
  </div>
</SettingsBoard>
`;
